### PR TITLE
Fix a memory leak in Routing Resource Graph Timing Parameter Annotation

### PR DESCRIPTION
**Pull request by tangxifan**
#### Description

Currently, there is a memory leak due to read an out-of-bound address in 

https://github.com/verilog-to-routing/vtr-verilog-to-routing/blob/a5b86e0c796b6ed6648291a8241d492232a7463e/vpr/src/route/rr_graph_generation/rr_graph_timing_params.cpp#L144-L146

This memory leak is captured by valgrind. See a detailed log here:

```
==958148== Invalid read of size 4
==958148==    at 0xC0B20B: add_rr_graph_C_from_switches() (rr_graph_timing_params.cpp:145)
==958148==    by 0xBAD06B: rr_graph_externals(std::vector<t_segment_inf, std::allocator<t_segment_inf> > const&, std::vector<t_segment_inf, std::allocator<t_segment_inf> > const&, std::vector<t_segment_inf, std::allocator<t_segment_inf> > const&, std::vector<t_segment_inf, std::allocator<t_segment_inf> > const&, e_base_cost_type, bool) (rr_graph.cpp:1200)
==958148==    by 0xC1D919: build_tileable_unidir_rr_graph(std::vector<t_physical_tile_type, std::allocator<t_physical_tile_type> > const&, DeviceGrid const&, t_chan_width const&, t_crr_opts const&, e_switch_block_type const&, int const&, e_switch_block_type const&, int const&, std::vector<t_segment_inf, std::allocator<t_segment_inf> > const&, int const&, int const&, float, float, e_base_cost_type const&, std::vector<t_direct_inf, std::allocator<t_direct_inf> > const&, bool const&, bool const&, bool const&, bool const&, bool const&, bool const&, int, int*, bool) (tileable_rr_graph_builder.cpp:320)
==958148==    by 0xBA8CFA: create_rr_graph(e_graph_type, std::vector<t_physical_tile_type, std::allocator<t_physical_tile_type> > const&, DeviceGrid const&, t_chan_width const&, t_det_routing_arch&, std::vector<t_segment_inf, std::allocator<t_segment_inf> > const&, t_router_opts const&, t_crr_opts const&, std::vector<t_direct_inf, std::allocator<t_direct_inf> > const&, int*, bool) (rr_graph.cpp:403)
==958148==    by 0x8C34EF: vpr_create_rr_graph(t_vpr_setup&, t_arch const&, int, bool) (vpr_api.cpp:1274)
==958148==    by 0x8C02DF: vpr_create_device(t_vpr_setup&, t_arch const&, bool) (vpr_api.cpp:579)
==958148==    by 0x8BFEF8: vpr_flow(t_vpr_setup&, t_arch&) (vpr_api.cpp:509)
==958148==    by 0x361BFD: vpr::vpr(int, char**) (vpr_main.cpp:57)
==958148==    by 0x362048: vpr::vpr_wrapper(int, char**) (vpr_main.cpp:100)
==958148==    by 0x361918: int std::__invoke_impl<int, int (*&)(int, char**), int, char**>(std::__invoke_other, int (*&)(int, char**), int&&, char**&&) (invoke.h:61)
==958148==    by 0x361698: std::enable_if<is_invocable_r_v<int, int (*&)(int, char**), int, char**>, int>::type std::__invoke_r<int, int (*&)(int, char**), int, char**>(int (*&)(int, char**), int&&, char**&&) (invoke.h:114)
==958148==    by 0x3614D8: std::_Function_handler<int (int, char**), int (*)(int, char**)>::_M_invoke(std::_Any_data const&, int&&, char**&&) (std_function.h:290)
==958148==  Address 0x7a903dc is 4 bytes before a block of size 32 alloc'd
==958148==    at 0x484A2F3: operator new[](unsigned long) (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==958148==    by 0xC0AC32: add_rr_graph_C_from_switches() (rr_graph_timing_params.cpp:41)
==958148==    by 0xBAD06B: rr_graph_externals(std::vector<t_segment_inf, std::allocator<t_segment_inf> > const&, std::vector<t_segment_inf, std::allocator<t_segment_inf> > const&, std::vector<t_segment_inf, std::allocator<t_segment_inf> > const&, std::vector<t_segment_inf, std::allocator<t_segment_inf> > const&, e_base_cost_type, bool) (rr_graph.cpp:1200)
==958148==    by 0xC1D919: build_tileable_unidir_rr_graph(std::vector<t_physical_tile_type, std::allocator<t_physical_tile_type> > const&, DeviceGrid const&, t_chan_width const&, t_crr_opts const&, e_switch_block_type const&, int const&, e_switch_block_type const&, int const&, std::vector<t_segment_inf, std::allocator<t_segment_inf> > const&, int const&, int const&, float, float, e_base_cost_type const&, std::vector<t_direct_inf, std::allocator<t_direct_inf> > const&, bool const&, bool const&, bool const&, bool const&, bool const&, bool const&, int, int*, bool) (tileable_rr_graph_builder.cpp:320)
==958148==    by 0xBA8CFA: create_rr_graph(e_graph_type, std::vector<t_physical_tile_type, std::allocator<t_physical_tile_type> > const&, DeviceGrid const&, t_chan_width const&, t_det_routing_arch&, std::vector<t_segment_inf, std::allocator<t_segment_inf> > const&, t_router_opts const&, t_crr_opts const&, std::vector<t_direct_inf, std::allocator<t_direct_inf> > const&, int*, bool) (rr_graph.cpp:403)
==958148==    by 0x8C34EF: vpr_create_rr_graph(t_vpr_setup&, t_arch const&, int, bool) (vpr_api.cpp:1274)
==958148==    by 0x8C02DF: vpr_create_device(t_vpr_setup&, t_arch const&, bool) (vpr_api.cpp:579)
==958148==    by 0x8BFEF8: vpr_flow(t_vpr_setup&, t_arch&) (vpr_api.cpp:509)
==958148==    by 0x361BFD: vpr::vpr(int, char**) (vpr_main.cpp:57)
==958148==    by 0x362048: vpr::vpr_wrapper(int, char**) (vpr_main.cpp:100)
==958148==    by 0x361918: int std::__invoke_impl<int, int (*&)(int, char**), int, char**>(std::__invoke_other, int (*&)(int, char**), int&&, char**&&) (invoke.h:61)
==958148==    by 0x361698: std::enable_if<is_invocable_r_v<int, int (*&)(int, char**), int, char**>, int>::type std::__invoke_r<int, int (*&)(int, char**), int, char**>(int (*&)(int, char**), int&&, char**&&) (invoke.h:114)
==958148== 
```

#### Motivation and Context

This memory leaks appears when there are connection blocks on boundary, which is enabled by the tileable routing resource graph generator when the ``perimeter_cb`` option is on.
See documentation here about the details: https://docs.verilogtorouting.org/en/latest/arch/reference/#cmdoption-arg-perimeter_cb

This results in the xlow(), ylow() could be 0 as routing tracks can start from ``x=0``.
Therefore, the ``is_block`` should not start from a offset of ``-1``. 
```
 for (isblock = iseg_low - 1; isblock <= iseg_high; isblock++) {
```
should be changed to 
```
 for (isblock = iseg_low; isblock <= iseg_high; isblock++) {
```

#### How Has This Been Tested?

This results minor QoR changes on number of CLBs (better than golden actually) on basic regression tests.
Per discussion on weekly meeting, relax the check bounds on VTR benchmarks from [90%, 110%] to [85%, 115%].

#### Types of changes

- [x] Bug fix (change which fixes an issue)
- [ ] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Checklist:

- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed

### CODE REVIEW

**Review by vaughnbetz — APPROVED** [review]
LGTM

**Comment by tangxifan** [comment]
@vaughnbetz I did look into the strong test failures due to Qor shift. It requires to extend the range from 85% downto 80%, which I think it is too much.
Since this memory corruption causes QoR golden shifted in a considerable number of testcase, I believe that updating golden is reasonable other than extend checking tolerance.

File: vpr/src/route/rr_graph_generation/rr_graph_timing_params.cpp [hunk 1]
```diff
@@ -141,7 +141,7 @@ void add_rr_graph_C_from_switches() {
                 cblock_counted[icblock] = false;
             }
 
-            for (isblock = iseg_low - 1; isblock <= iseg_high; isblock++) {
+            for (isblock = iseg_low; isblock <= iseg_high; isblock++) {
```
Comment (AlexandreSinger) [inline vpr/src/route/rr_graph_generation/rr_graph_timing_params.cpp]: I also agree with Amir's comment here. We have Valgrind tests in VTR for a few basic architectures and they do not show this behavior... (at least not in the CI).

I am wondering if this is specific to tileable architectures.

**Comment by AmirhosseinPoolad** [comment]
~~I took another look at the code and I don't think the indexing is an issue (removed my previous comment), but~~ regenerating \*all\* golden results is overkill. Let's only regenerate the failing tests and on a case by case basis.

EDIT: Sorry for constantly changing my mind here, this is a complicated function to analyze. I'm not completely convinced that this is a safe fix. From my understanding, isblock (switch block index) is started from `iseg_low - 1` because a wire can be driven by the switch block that's -1 behind it. This is a VPR coordinate system quirk essentially.

Drawing

Excuse my terrible sketch. The CHANX segment at (1,2) can be driven by the switch block at (0,2) which is why the -1 is in there. The CHANX segment at (0,2) does not have any switch blocks on its left side. I believe default VPR does not add segments at X=0 and Y=0 so this problem doesn't exist, but this is different in tillable architectures. Overall a proper fix for this should *not* change the results for any non-tillable architecture.

**Comment by tangxifan** [comment]
@AmirhosseinPoolad No worries. It is good to have you and @AlexandreSinger double check on this PR :)

In VPR's regular routing architecture, there are no routing tracks in (x=0, y=2). 
In the for-loop here, you may get 
- CHANY at node_xlow, node_ylow) = (0,2) 
- but **not** CHANX (node_xlow, node_ylow) = (0,2) 

So, you will not get invalid address when applying -1 on the index.

In tileable routing architecture, when the option ``perimeter_cb`` is ``true``, there are routing tracks at (x=0, y=2). 
In the for-loop here, you may get 
- CHANY at node_xlow, node_ylow) = (0,2) 
- but **also** CHANX (node_xlow, node_ylow) = (0,2) 

Then, you may get some data at (x=-1, y=2)

 I can patch it like:

```
int iseg_low_start_idx = iseg_low - 1;
if (iseg_low == 0) {
  iseg_low_start_idx = 0;
}
for (isblock = iseg_low_start_idx; isblock <= iseg_high; isblock++) { 
     rr_node_C[inode] += buffer_Cin[isblock]; /* Biggest buf Cin at loc */ 
     buffer_Cin[isblock] = 0.;
}
```

**Comment by tangxifan** [comment]
@AmirhosseinPoolad @AlexandreSinger Here is the solution I propose here:

- Modify the for-loop starting index to be conditional on the boundary values
- Add a new testcase for valgrind test, where the ``perimeter_cb`` is on.
- Only update golden tests where we see changes

**Comment by AmirhosseinPoolad** [comment]
I think this is a good solution. Thanks @tangxifan!

**Comment by tangxifan** [comment]
@vaughnbetz @AmirhosseinPoolad @AlexandreSinger Please kindly check the revision of this PR. 

- No Qor changes required now. 
- All tests passed except the tests failed on master already.
- New tests added to valgrind test set.

It is ready for merge. Waiting for your final comments.

**Review by AlexandreSinger — APPROVED** [review]
LGTM.

The two failing testcases are irrelevant to this change. CI has been acting strange today. If things are still messy on Monday, I will give it a look.

The runtime for the valgrind test increased from 1.5 mins to 10 mins; but to be honest I am ok with the increase in run time. We are now testing 4 different circuits and two different architectures which will help find Valgrind errors. I think the multithreading is also causing this to slow down, so some future CI improvements may be needed. We should also add a weekly test that tests much larger Valgrind tests. This is not for this PR, but just thinking out loud to prevent issues like this in the future.

**Comment by vaughnbetz** [comment]
Nice catch @AmirhosseinPoolad !
